pkg/utils/backoff/backoff_set.go

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backoff
+
+import (
+	"sync"
+	"time"
+
+	"sigs.k8s.io/kwok/pkg/utils/wait"
```

Comment:
It feels like dependence is reversed

Reply:
@wzshiming  Yes. This will cause an import cycle if we put `RetryQueue` in `wait` package. Any suggestion about the layout? Or if it's not feasible, I can also refactor the code based on your suggestions.

the current implementation is  based on your comments:
https://github.com/kubernetes-sigs/kwok/issues/816#issuecomment-1869461211

Reply:
https://github.com/kubernetes-sigs/kwok/pull/902/files#diff-b0dafdbd30f76f38d1f10f9abfdbb4adda9d8eaa2d674ef18134ac7d50c0bdf7R113-R139

In this PR, I have tried to replace Normal Queue and Failed Queue with Priority Queue. It works pretty well with the proportional processing of events.


then the number of failures of the operation is recorded instead of the number of seconds of the last sleep, and the number of seconds of the next sleep is calculated based on the number of failures each time.

Reply:
Thanks @wzshiming.  Will dig into #902 and digest what you said here.

Reply:
Hi, @wzshiming. Do you mean once a new event of a resource **type**(not a **individual** resource) comes in, the **entire** failed queue should be canceled? So there is no need to maintain a failed queue, right?

Reply:
Yes and the weight of the failure queue should be the lowest

Reply:
In the case of a single queue, normal will be directly blocked once a large batch of updates fail. Although separating the normal and failure queues alleviates this issue to some extent, it still results in the blocking of the normal. To address this, we enforce a normal queue using the weighted queue.

Reply:
> The overall design is to prioritize new events and **failures are only processed when there are no new events**.

@wzshiming  Maybe I misunderstood what you meant here.

As per the implementation of the weighted queue, If we put the failed tasks into the queue with the lowest weight (`1`), **they still have a chance to be processed**  if new events are coming **but just with a relatively low proportion**. This is somewhat contradictory to what you said (quoted) 

Could you please further clarify it? Thanks!

Reply:
Hi, @wzshiming 

The following are my understanding for the retry feature:

1. replace the delay queue with weight delay queue
2. for a normal(fresh) stage job, push it in the weighted delay queue with weight `0` (highest priority)
3. when *`playStage()`* fails in *`playStageWorker`*, we repush it immediately to the weight delay queue with weight `1` as well as a **backoff delay**. The backoff period should be calculated  based on the **failed times** (according to your new comments). Thus we also need a dediated map to track the failed times of each job. My question here:
   
    - **do we need a dedicated struct for this map? (since the associated  backoff settings(cap, factor, jitter) also need to be recorded.)**
    - If above if true, **where should I put the instance of this struct?** Should I put it directly under the Controller (`PodController`, `NodeController`) as the `delayQueueMapping` one?
    
Please correct me if there is any misunderstanding. Thanks again for your great patience! ;-)

Reply:
Thank you, in my opinion, maybe keep track of the number of failures. then calculate the next time when failed, such as `jitter(min(100ms * (2 ** failed_number), maximal_backoff), jitter_factor)`.

Reply:
I have an idea that:

 Using Job  struct  *`resourceStageJob`* itself to record the failed count instead of using an additional dedicated map to track the failed counts.

I think there may several drawbacks if using an additional  map:

Since we already have an associated map struct `delayQueueMapping,` maintaining too many scattered map structs around the queue is not a good idea. For example, when we cancel a job, we also need to remove the associated item from that map. https://github.com/kubernetes-sigs/kwok/blob/main/pkg/kwok/controllers/pod_controller.go#L252

Also, if we add a failed job to the queue, we need to add them both in `delayQueueMapping` and the failed count map.


In terms of implementation, I want to add  a field `RetryCount`  to *`resourceStageJob`* :

```go
type resourceStageJob[T any] struct {
	Resource T
	Stage    *LifecycleStage
	Key      string
	
        RetryCount int
}
```

When a job is pushed to the queue for the first time, the `RetryCount` should be zero. Once  *`playStage()`* fails in *`playStageWorker` ,* we incease the failed count and repush it to the queue with weight `1` and a delay time, which is calculated based on the increased retry times.


What do you think?

Reply:
Good idea.

Note that this requires using a pointer and initializing it, otherwise it won't match the original value
``` golang
RetryCount *int
```